cpp/src/replace/clamp.cu

```diff
@@ -0,0 +1,385 @@
+            if(lo_validity_pair.second and
+                (element_validity_pair.first < lo_validity_pair.first)) {
+                return thrust::get<1>(scalar_tuple).first;
+            } else if(hi_validity_pair.second and
+                (element_validity_pair.first > hi_validity_pair.first)) {
+                return thrust::get<3>(scalar_tuple).first;
+            }
+        }
+
+        return element_validity_pair.first;
+    };
+
+    if (input.has_nulls()) {
+        auto input_pair_iterator = make_pair_iterator<T, true>(*input_device_view);
+        apply_transform<T>(input_pair_iterator, input_pair_iterator+input.size(),
+                           scalar_zip_itr, *output_device_view, trans, stream);
+    } else {
+        auto input_pair_iterator = make_pair_iterator<T, false>(*input_device_view);
+        apply_transform<T>(input_pair_iterator, input_pair_iterator+input.size(),
+                           scalar_zip_itr, *output_device_view, trans, stream);
+    }
+
+    return output;
+}
+
+template <typename T, typename ScalarIterator>
+std::enable_if_t<std::is_same<T, string_view>::value, std::unique_ptr<cudf::column>>
+clamper (column_view const& input,
+         ScalarIterator const& lo_itr,
+         ScalarIterator const& lo_replace_itr,
+         ScalarIterator const& hi_itr,
+         ScalarIterator const& hi_replace_itr,
+         rmm::mr::device_memory_resource* mr,
+         cudaStream_t stream) {
+
+    return clamp_string_column (input, lo_itr, lo_replace_itr,
+                                hi_itr, hi_replace_itr, mr, stream);
+}
+
+} //namespace
+
+template<typename T, typename ScalarIterator>
+std::unique_ptr<column> clamp(column_view const& input,
+                              ScalarIterator const& lo_itr,
+                              ScalarIterator const& lo_replace_itr,
+                              ScalarIterator const& hi_itr,
+                              ScalarIterator const& hi_replace_itr,
+                              rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource(),
+                              cudaStream_t stream = 0) {
+    return clamper<T>(input, lo_itr, lo_replace_itr, hi_itr, hi_replace_itr, mr, stream);
+}
+
+struct dispatch_clamp {
+    template<typename T>
+    std::unique_ptr<column> operator ()(column_view const& input,
+                scalar const& lo,
+                scalar const& lo_replace,
+                scalar const& hi,
+                scalar const& hi_replace,
+                rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource(),
+                cudaStream_t stream = 0) {
+        
+        auto lo_itr = make_pair_iterator<T>(lo);
+        auto hi_itr = make_pair_iterator<T>(hi);
+        auto lo_replace_itr = make_pair_iterator<T>(lo_replace);
+        auto hi_replace_itr = make_pair_iterator<T>(hi_replace);
+
+
+        return clamp<T>(input, lo_itr, lo_replace_itr,
```

Comment:
`clamp<T>` seems to just directly pass through to `clamper<T>`. Can you eliminate a layer (`clamp<T>`) here?

Reply:
Wanted to leave an API in details if anyone was interested in using clamp with Iterators and keep implementation in nameless namespace.